File: android/src/main/java/com/rnvisioncameramlkit/BarcodeScanningPlugin.kt
package com.rnvisioncameramlkit

import android.graphics.Point
import android.graphics.Rect
import android.media.Image
import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.barcode.BarcodeScanner
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy
import com.rnvisioncameramlkit.utils.Logger

/**
 * Barcode Scanning Frame Processor Plugin
 *
 * Performs on-device barcode scanning using Google ML Kit
 * Supports 1D and 2D formats with structured data extraction
 */
class BarcodeScanningPlugin(
    proxy: VisionCameraProxy,
    options: Map<String, Any>?
) : FrameProcessorPlugin() {

    private var scanner: BarcodeScanner

    init {
        Logger.info("Initializing barcode scanner")

        val formats = options?.get("formats") as? List<*>
        val scannerOptions = if (formats != null && formats.isNotEmpty()) {
            val barcodeFormats = formats.mapNotNull { formatString ->
                parseBarcode Format(formatString.toString())
            }

            if (barcodeFormats.isEmpty()) {
                Logger.warn("No valid barcode formats specified, scanning all formats")
                BarcodeScannerOptions.Builder()
                    .setBarcodeFormats(Barcode.FORMAT_ALL_FORMATS)
                    .build()
            } else {
                Logger.info("Scanning specific formats: $formats")
                BarcodeScannerOptions.Builder()
                    .setBarcodeFormats(
                        barcodeFormats.first(),
                        *barcodeFormats.drop(1).toIntArray()
                    )
                    .build()
            }
        } else {
            Logger.info("Scanning all barcode formats")
            BarcodeScannerOptions.Builder()
                .setBarcodeFormats(Barcode.FORMAT_ALL_FORMATS)
                .build()
        }

        scanner = BarcodeScanning.getClient(scannerOptions)
        Logger.info("Barcode scanner initialized successfully")
    }

    override fun callback(frame: Frame, arguments: Map<String, Any>?): Any? {
        val startTime = System.currentTimeMillis()

        try {
            val mediaImage: Image = frame.image
            val image = InputImage.fromMediaImage(
                mediaImage,
                frame.imageProxy.imageInfo.rotationDegrees
            )

            Logger.debug("Processing frame: ${frame.width}x${frame.height}, rotation: ${frame.imageProxy.imageInfo.rotationDegrees}")

            val task: Task<List<Barcode>> = scanner.process(image)
            val barcodes: List<Barcode> = Tasks.await(task)

            val processingTime = System.currentTimeMillis() - startTime
            Logger.performance("Barcode scanning processing", processingTime)

            if (barcodes.isEmpty()) {
                Logger.debug("No barcodes detected in frame")
                return null
            }

            Logger.debug("Barcodes detected: ${barcodes.size} barcode(s)")

            val result = WritableNativeMap().apply {
                putArray("barcodes", processBarcodes(barcodes))
            }

            return result.toHashMap()

        } catch (e: Exception) {
            val processingTime = System.currentTimeMillis() - startTime
            Logger.error("Error during barcode scanning", e)
            Logger.performance("Barcode scanning processing (error)", processingTime)
            return null
        }
    }

    companion object {
        /**
         * Parse barcode format string to ML Kit format constant
         */
        private fun parseBarcodeFormat(format: String): Int? {
            return when (format.lowercase()) {
                "codabar" -> Barcode.FORMAT_CODABAR
                "code39" -> Barcode.FORMAT_CODE_39
                "code93" -> Barcode.FORMAT_CODE_93
                "code128" -> Barcode.FORMAT_CODE_128
                "ean8" -> Barcode.FORMAT_EAN_8
                "ean13" -> Barcode.FORMAT_EAN_13
                "itf" -> Barcode.FORMAT_ITF
                "upca" -> Barcode.FORMAT_UPC_A
                "upce" -> Barcode.FORMAT_UPC_E
                "aztec" -> Barcode.FORMAT_AZTEC
                "datamatrix" -> Barcode.FORMAT_DATA_MATRIX
                "pdf417" -> Barcode.FORMAT_PDF417
                "qrcode" -> Barcode.FORMAT_QR_CODE
                else -> {
                    Logger.warn("Unknown barcode format: $format")
                    null
                }
            }
        }

        /**
         * Convert ML Kit barcode format to string
         */
        private fun barcodeFormatToString(format: Int): String {
            return when (format) {
                Barcode.FORMAT_CODABAR -> "codabar"
                Barcode.FORMAT_CODE_39 -> "code39"
                Barcode.FORMAT_CODE_93 -> "code93"
                Barcode.FORMAT_CODE_128 -> "code128"
                Barcode.FORMAT_EAN_8 -> "ean8"
                Barcode.FORMAT_EAN_13 -> "ean13"
                Barcode.FORMAT_ITF -> "itf"
                Barcode.FORMAT_UPC_A -> "upca"
                Barcode.FORMAT_UPC_E -> "upce"
                Barcode.FORMAT_AZTEC -> "aztec"
                Barcode.FORMAT_DATA_MATRIX -> "dataMatrix"
                Barcode.FORMAT_PDF417 -> "pdf417"
                Barcode.FORMAT_QR_CODE -> "qrCode"
                else -> "unknown"
            }
        }

        /**
         * Convert ML Kit value type to string
         */
        private fun valueTypeToString(valueType: Int): String {
            return when (valueType) {
                Barcode.TYPE_TEXT -> "text"
                Barcode.TYPE_URL -> "url"
                Barcode.TYPE_EMAIL -> "email"
                Barcode.TYPE_PHONE -> "phone"
                Barcode.TYPE_SMS -> "sms"
                Barcode.TYPE_WIFI -> "wifi"
                Barcode.TYPE_GEO -> "geo"
                Barcode.TYPE_CONTACT_INFO -> "contact"
                Barcode.TYPE_CALENDAR_EVENT -> "calendarEvent"
                Barcode.TYPE_DRIVER_LICENSE -> "driverLicense"
                else -> "unknown"
            }
        }

        /**
         * Process barcodes into React Native compatible format
         */
        private fun processBarcodes(barcodes: List<Barcode>): WritableNativeArray {
            val barcodeArray = WritableNativeArray()

            for (barcode in barcodes) {
                val barcodeMap = WritableNativeMap().apply {
                    putString("rawValue", barcode.rawValue ?: "")
                    putString("displayValue", barcode.displayValue ?: "")
                    putString("format", barcodeFormatToString(barcode.format))
                    putString("valueType", valueTypeToString(barcode.valueType))

                    // Bounding box and corner points
                    putMap("frame", processRect(barcode.boundingBox))
                    putArray("cornerPoints", processCornerPoints(barcode.cornerPoints))

                    // Structured data based on type
                    when (barcode.valueType) {
                        Barcode.TYPE_WIFI -> {
                            barcode.wifi?.let { wifi ->
                                putMap("wifi", WritableNativeMap().apply {
                                    putString("ssid", wifi.ssid ?: "")
                                    putString("password", wifi.password ?: "")
                                    putString("encryptionType", when (wifi.encryptionType) {
                                        Barcode.WiFi.TYPE_OPEN -> "open"
                                        Barcode.WiFi.TYPE_WPA -> "wpa"
                                        Barcode.WiFi.TYPE_WEP -> "wep"
                                        else -> "unknown"
                                    })
                                })
                            }
                        }
                        Barcode.TYPE_URL -> {
                            barcode.url?.let { url ->
                                putString("url", url.url ?: "")
                            }
                        }
                        Barcode.TYPE_EMAIL -> {
                            barcode.email?.let { email ->
                                putString("email", email.address ?: "")
                            }
                        }
                        Barcode.TYPE_PHONE -> {
                            barcode.phone?.let { phone ->
                                putString("phone", phone.number ?: "")
                            }
                        }
                        Barcode.TYPE_SMS -> {
                            barcode.sms?.let { sms ->
                                putMap("sms", WritableNativeMap().apply {
                                    putString("phoneNumber", sms.phoneNumber ?: "")
                                    putString("message", sms.message ?: "")
                                })
                            }
                        }
                        Barcode.TYPE_GEO -> {
                            barcode.geoPoint?.let { geo ->
                                putMap("geo", WritableNativeMap().apply {
                                    putDouble("latitude", geo.lat)
                                    putDouble("longitude", geo.lng)
                                })
                            }
                        }
                        Barcode.TYPE_CONTACT_INFO -> {
                            barcode.contactInfo?.let { contact ->
                                putMap("contact", WritableNativeMap().apply {
                                    contact.name?.let { name ->
                                        putString("name", "${name.first ?: ""} ${name.last ?: ""}".trim())
                                    }
                                    putString("organization", contact.organization ?: "")

                                    contact.phones?.let { phones ->
                                        val phoneArray = WritableNativeArray()
                                        phones.forEach { phone ->
                                            phoneArray.pushString(phone.number ?: "")
                                        }
                                        putArray("phones", phoneArray)
                                    }

                                    contact.emails?.let { emails ->
                                        val emailArray = WritableNativeArray()
                                        emails.forEach { email ->
                                            emailArray.pushString(email.address ?: "")
                                        }
                                        putArray("emails", emailArray)
                                    }

                                    contact.urls?.let { urls ->
                                        val urlArray = WritableNativeArray()
                                        urls.forEach { url ->
                                            urlArray.pushString(url ?: "")
                                        }
                                        putArray("urls", urlArray)
                                    }

                                    contact.addresses?.let { addresses ->
                                        val addressArray = WritableNativeArray()
                                        addresses.forEach { address ->
                                            val addressLines = listOfNotNull(
                                                address.addressLines?.joinToString(", ")
                                            )
                                            addressArray.pushString(addressLines.joinToString(" "))
                                        }
                                        putArray("addresses", addressArray)
                                    }
                                })
                            }
                        }
                        Barcode.TYPE_CALENDAR_EVENT -> {
                            barcode.calendarEvent?.let { event ->
                                putMap("calendarEvent", WritableNativeMap().apply {
                                    putString("summary", event.summary ?: "")
                                    putString("description", event.description ?: "")
                                    putString("location", event.location ?: "")
                                    event.start?.let { start ->
                                        putString("start", start.rawValue ?: "")
                                    }
                                    event.end?.let { end ->
                                        putString("end", end.rawValue ?: "")
                                    }
                                })
                            }
                        }
                        Barcode.TYPE_DRIVER_LICENSE -> {
                            barcode.driverLicense?.let { license ->
                                putMap("driverLicense", WritableNativeMap().apply {
                                    putString("documentType", license.documentType ?: "")
                                    putString("firstName", license.firstName ?: "")
                                    putString("lastName", license.lastName ?: "")
                                    putString("gender", license.gender ?: "")
                                    putString("addressStreet", license.addressStreet ?: "")
                                    putString("addressCity", license.addressCity ?: "")
                                    putString("addressState", license.addressState ?: "")
                                    putString("addressZip", license.addressZip ?: "")
                                    putString("licenseNumber", license.licenseNumber ?: "")
                                    putString("issueDate", license.issueDate ?: "")
                                    putString("expiryDate", license.expiryDate ?: "")
                                    putString("birthDate", license.birthDate ?: "")
                                    putString("issuingCountry", license.issuingCountry ?: "")
                                })
                            }
                        }
                    }
                }
                barcodeArray.pushMap(barcodeMap)
            }

            return barcodeArray
        }

        /**
         * Convert Android Rect to React Native format
         */
        private fun processRect(boundingBox: Rect?): WritableNativeMap {
            val rectMap = WritableNativeMap()

            boundingBox?.let { box ->
                rectMap.putDouble("x", box.exactCenterX().toDouble())
                rectMap.putDouble("y", box.exactCenterY().toDouble())
                rectMap.putInt("width", box.width())
                rectMap.putInt("height", box.height())
            }

            return rectMap
        }

        /**
         * Convert Android corner points to React Native format
         */
        private fun processCornerPoints(cornerPoints: Array<Point>?): WritableNativeArray {
            val pointsArray = WritableNativeArray()

            cornerPoints?.forEach { point ->
                val pointMap = WritableNativeMap().apply {
                    putInt("x", point.x)
                    putInt("y", point.y)
                }
                pointsArray.pushMap(pointMap)
            }

            return pointsArray
        }
    }
}

File: ios/BarcodeScanningPlugin.swift
//
//  BarcodeScanningPlugin.swift
//  react-native-vision-camera-ml-kit
//

import Foundation
import VisionCamera
import MLKitVision
import MLKitCommon
import MLKitBarcodeScanning
import CoreImage
import CoreMedia

/// Normalized scan region (0..1) in the upright frame, or in a cover-fitted
/// preview view when viewport dims are set
struct ScanRegionSpec {
    let left: CGFloat
    let top: CGFloat
    let width: CGFloat
    let height: CGFloat
    let viewportWidth: CGFloat?
    let viewportHeight: CGFloat?

    static func parse(_ value: Any?) -> ScanRegionSpec? {
        guard let map = value as? [AnyHashable: Any],
              let left = (map["left"] as? NSNumber)?.doubleValue,
              let top = (map["top"] as? NSNumber)?.doubleValue,
              let width = (map["width"] as? NSNumber)?.doubleValue,
              let height = (map["height"] as? NSNumber)?.doubleValue else {
            return nil
        }
        if width <= 0 || height <= 0 {
            Logger.warn("Invalid scanRegion: width and height must be positive. Got width=\(width), height=\(height)")
            return nil
        }
        return ScanRegionSpec(
            left: CGFloat(left),
            top: CGFloat(top),
            width: CGFloat(width),
            height: CGFloat(height),
            viewportWidth: (map["viewportWidth"] as? NSNumber).map { CGFloat($0.doubleValue) },
            viewportHeight: (map["viewportHeight"] as? NSNumber).map { CGFloat($0.doubleValue) }
        )
    }
}

@objc(BarcodeScanningPlugin)
public class BarcodeScanningPlugin: FrameProcessorPlugin {

    private var scanner: BarcodeScanner!
    private var detectInvertedBarcodes: Bool = false
    private var tryRotations: Bool = true
    private var scanRegion: ScanRegionSpec?
    private var warnedInvalidRegion = false

    public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]! = [:]) {
        super.init(proxy: proxy, options: options)

        Logger.info("Initializing barcode scanner")

        // Extract options
        detectInvertedBarcodes = options["detectInvertedBarcodes"] as? Bool ?? false
        if detectInvertedBarcodes {
            Logger.warn("⚠️ Inverted barcode detection ENABLED - adds processing time per frame when no barcodes found. Only enable if you specifically need white-on-black barcodes.")
        }

        tryRotations = options["tryRotations"] as? Bool ?? true
        if !tryRotations {
            Logger.info("90 degree rotation attempts DISABLED - only using current camera rotation")
        }

        scanRegion = ScanRegionSpec.parse(options["scanRegion"])
        if let region = scanRegion {
            Logger.info("Scan region enabled: \(region) — decoding restricted to this area")
        }

        // Parse formats
        let formats = options["formats"] as? [String]
        let scannerOptions = createScannerOptions(formats: formats)

        scanner = BarcodeScanner.barcodeScanner(options: scannerOptions)
        Logger.info("Barcode scanner initialized successfully")
    }

    deinit {
        // Clean up ML Kit resources when plugin is deallocated
        // Swift ARC will handle the deallocation, but we log for debugging
        Logger.debug("BarcodeScanningPlugin deallocating - ML Kit scanner resources will be freed")
        // Note: ML Kit resources are automatically freed by ARC when scanner is deallocated
    }

    private func createScannerOptions(formats: [String]?) -> BarcodeScannerOptions {
        guard let formats = formats, !formats.isEmpty else {
            Logger.info("No format filter specified, scanning all barcode formats")
            return BarcodeScannerOptions(formats: .all)
        }

        Logger.debug("Parsing \(formats.count) barcode format(s) from options")

        var combinedFormats: BarcodeFormat = []
        for formatString in formats {
            if let parsedFormat = parseBarcodeFormat(formatString) {
                combinedFormats.insert(parsedFormat)
                Logger.debug("Successfully parsed format: '\(formatString)'")
            } else {
                Logger.error("FAILED to parse barcode format: '\(formatString)'")
            }
        }

        if combinedFormats.isEmpty {
            Logger.error("No valid barcode formats could be parsed! Falling back to all formats")
            return BarcodeScannerOptions(formats: .all)
        }

        Logger.info("Scanning barcode format(s) with combined mask")
        return BarcodeScannerOptions(formats: combinedFormats)
    }

    public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
        let startTime = Date()

        do {
            let orientation = frame.orientation
            let baseOrientation = getOrientation(orientation: orientation)

            if Logger.isDebugEnabled() {
                Logger.debug("Processing frame: \(frame.width)x\(frame.height), orientation: \(orientation.rawValue)")
            }

            // Build list of orientations to try
            let orientations: [UIImage.Orientation] = tryRotations
                ? [baseOrientation, rotateOrientation90(baseOrientation)]
                : [baseOrientation]

            var barcodes: [Barcode] = []

            // Clone the camera buffer to UIImage to release the original buffer immediately
            // This prevents buffer exhaustion issues when ML Kit processing takes longer than camera frame rate
            guard let baseImage = ImageUtils.imageFromSampleBuffer(frame.buffer) else {
                Logger.error("Failed to clone sample buffer to image")
                return nil
            }

            let region = ScanRegionSpec.parse(arguments?["scanRegion"]) ?? scanRegion

            // Crop to the scan region so ML Kit never sees barcodes outside it
            var workingImage = baseImage
            var cropOffsetX: CGFloat = 0
            var cropOffsetY: CGFloat = 0
            if let region = region {
                let rawWidth = baseImage.size.width
                let rawHeight = baseImage.size.height
                guard let crop = Self.computeCropRect(
                    region: region,
                    rawWidth: rawWidth,
                    rawHeight: rawHeight,
                    rotationDegrees: Self.rotationDegrees(for: baseOrientation)
                ) else {
                    if !warnedInvalidRegion {
                        warnedInvalidRegion = true
                        Logger.warn("scanRegion resolves to an empty area — skipping decode. Region: \(region), frame: \(rawWidth)x\(rawHeight)")
                    }
                    return nil
                }
                if crop.minX > 0 || crop.minY > 0 || crop.width < rawWidth || crop.height < rawHeight {
                    guard let cgImage = baseImage.cgImage, let croppedCG = cgImage.cropping(to: crop) else {
                        Logger.error("Failed to crop frame to scan region")
                        return nil
                    }
                    workingImage = UIImage(cgImage: croppedCG, scale: 1.0, orientation: .up)
                    cropOffsetX = crop.minX
                    cropOffsetY = crop.minY
                }
            }

            // 1. Try normal image at current orientation
            let visionImage = VisionImage(image: workingImage)
            visionImage.orientation = orientations[0]
            barcodes = try scanner.results(in: visionImage)

            if !barcodes.isEmpty {
                if Logger.isDebugEnabled() {
                    Logger.debug("Found \(barcodes.count) barcode(s) at orientation \(orientations[0].rawValue)")
                }
            } else if tryRotations && orientations.count > 1 {
                // 2. Try normal image at 90 degree rotation
                if Logger.isDebugEnabled() {
                    Logger.debug("No barcodes at orientation \(orientations[0].rawValue), trying \(orientations[1].rawValue)")
                }
                visionImage.orientation = orientations[1]
                barcodes = try scanner.results(in: visionImage)

                if !barcodes.isEmpty && Logger.isDebugEnabled() {
                    Logger.debug("Found \(barcodes.count) barcode(s) at orientation \(orientations[1].rawValue)")
                }
            }

            // 3. If no barcodes found and inverted detection is enabled, try inverted images
            if barcodes.isEmpty && detectInvertedBarcodes {
                if Logger.isDebugEnabled() {
                    Logger.debug("No barcodes in normal image, attempting inverted image scan...")
                }

                let invertStartTime = Date()

                if let workingCG = workingImage.cgImage,
                   let invertedCI = ImageUtils.invertImage(CIImage(cgImage: workingCG)),
                   let invertedImage = ImageUtils.uiImageFromCIImage(invertedCI) {
                    // Try inverted at current orientation
                    let invertedVisionImage = VisionImage(image: invertedImage)
                    invertedVisionImage.orientation = orientations[0]
                    barcodes = try scanner.results(in: invertedVisionImage)

                    if !barcodes.isEmpty {
                        if Logger.isDebugEnabled() {
                            Logger.debug("Found \(barcodes.count) barcode(s) in inverted image at orientation \(orientations[0].rawValue)")
                        }
                    } else if tryRotations && orientations.count > 1 {
                        // Try inverted at 90 degree rotation
                        if Logger.isDebugEnabled() {
                            Logger.debug("No barcodes in inverted at \(orientations[0].rawValue), trying \(orientations[1].rawValue)")
                        }
                        invertedVisionImage.orientation = orientations[1]
                        barcodes = try scanner.results(in: invertedVisionImage)

                        if !barcodes.isEmpty && Logger.isDebugEnabled() {
                            Logger.debug("Found \(barcodes.count) barcode(s) in inverted image at orientation \(orientations[1].rawValue)")
                        }
                    }
                }

                let invertTime = Int64(Date().timeIntervalSince(invertStartTime) * 1000)
                Logger.performance("Inverted image scan", durationMs: invertTime)
            }

            let processingTime = Int64(Date().timeIntervalSince(startTime) * 1000)
            Logger.performance("Barcode scanning processing", durationMs: processingTime)

            if barcodes.isEmpty {
                if Logger.isDebugEnabled() {
                    Logger.debug("No barcodes detected in frame (tried all orientations)")
                }
                return nil
            }

            if Logger.isDebugEnabled() {
                Logger.debug("Barcodes detected: \(barcodes.count) barcode(s)")
            }

            // iOS ML Kit returns coordinates in the source image space, so the
            // crop offset shifts them back to the full frame regardless of rotation
            let result: [String: Any] = ["barcodes": processBarcodes(barcodes, offsetX: cropOffsetX, offsetY: cropOffsetY)]
            return result

        } catch {
            let processingTime = Int64(Date().timeIntervalSince(startTime) * 1000)
            Logger.error("Exception during barcode scanning: \(error.localizedDescription)")
            Logger.performance("Barcode scanning processing (error)", durationMs: processingTime)
            return nil
        }
    }
    
    // MARK: - Image Processing
    
    /// Rotate orientation by 90 degrees clockwise
    private func rotateOrientation90(_ orientation: UIImage.Orientation) -> UIImage.Orientation {
        switch orientation {
        case .up: return .right
        case .right: return .down
        case .down: return .left
        case .left: return .up
        case .upMirrored: return .rightMirrored
        case .rightMirrored: return .downMirrored
        case .downMirrored: return .leftMirrored
        case .leftMirrored: return .upMirrored
        @unknown default: return .right
        }
    }
    
    /// Degrees the raw buffer must be rotated clockwise to appear upright
    private static func rotationDegrees(for orientation: UIImage.Orientation) -> Int {
        switch orientation {
        case .up, .upMirrored: return 0
        case .right, .rightMirrored: return 90
        case .down, .downMirrored: return 180
        case .left, .leftMirrored: return 270
        @unknown default: return 0
        }
    }

    /// Raw-buffer crop rect for a scan region defined in the upright frame
    /// (or in a cover-fitted preview view when viewport dims are present).
    /// Returns nil when the region resolves to an empty area.
    static func computeCropRect(
        region: ScanRegionSpec,
        rawWidth: CGFloat,
        rawHeight: CGFloat,
        rotationDegrees: Int
    ) -> CGRect? {
        let rot = ((rotationDegrees % 360) + 360) % 360
        let uprightW = (rot == 90 || rot == 270) ? rawHeight : rawWidth
        let uprightH = (rot == 90 || rot == 270) ? rawWidth : rawHeight

        var l = region.left
        var t = region.top
        var r = region.left + region.width
        var b = region.top + region.height

        if let vw = region.viewportWidth, let vh = region.viewportHeight, vw > 0, vh > 0 {
            // Map view-relative → frame-relative through the cover-fit crop
            let scale = max(vw / uprightW, vh / uprightH)
            let padX = (uprightW * scale - vw) / 2
            let padY = (uprightH * scale - vh) / 2
            l = (l * vw + padX) / scale / uprightW
            t = (t * vh + padY) / scale / uprightH
            r = (r * vw + padX) / scale / uprightW
            b = (b * vh + padY) / scale / uprightH
        }

        l = min(max(l, 0), 1)
        t = min(max(t, 0), 1)
        r = min(max(r, 0), 1)
        b = min(max(b, 0), 1)
        if r <= l || b <= t {
            return nil
        }

        // Upright-normalized → raw buffer space (inverse of the display rotation)
        let rawL: CGFloat
        let rawT: CGFloat
        let rawR: CGFloat
        let rawB: CGFloat
        switch rot {
        case 90:
            rawL = t; rawT = 1 - r; rawR = b; rawB = 1 - l
        case 180:
            rawL = 1 - r; rawT = 1 - b; rawR = 1 - l; rawB = 1 - t
        case 270:
            rawL = 1 - b; rawT = l; rawR = 1 - t; rawB = r
        default:
            rawL = l; rawT = t; rawR = r; rawB = b
        }

        let left = min(max(floor(rawL * rawWidth), 0), rawWidth - 1)
        let top = min(max(floor(rawT * rawHeight), 0), rawHeight - 1)
        let right = min(max(ceil(rawR * rawWidth), left + 1), rawWidth)
        let bottom = min(max(ceil(rawB * rawHeight), top + 1), rawHeight)
        return CGRect(x: left, y: top, width: right - left, height: bottom - top)
    }

    // MARK: - Orientation Mapping

    private func getOrientation(orientation: UIImage.Orientation) -> UIImage.Orientation {
        switch orientation {
        case .up:
            return .up
        case .left:
            return .right  // Swap left and right
        case .down:
            return .down
        case .right:
            return .left   // Swap left and right
        default:
            return .up
        }
    }

    // MARK: - Format Parsing

    private func parseBarcodeFormat(_ format: String) -> BarcodeFormat? {
        switch format.lowercased() {
        case "codabar": return .codaBar
        case "code39": return .code39
        case "code93": return .code93
        case "code128": return .code128
        case "ean8": return .EAN8
        case "ean13": return .EAN13
        case "itf": return .ITF
        case "upca": return .UPCA
        case "upce": return .UPCE
        case "aztec": return .aztec
        case "datamatrix": return .dataMatrix
        case "pdf417": return .PDF417
        case "qrcode": return .qrCode
        default:
            Logger.warn("Unknown barcode format: \(format)")
            return nil
        }
    }

    private func barcodeFormatToString(_ format: BarcodeFormat) -> String {
        switch format {
        case .codaBar: return "codabar"
        case .code39: return "code39"
        case .code93: return "code93"
        case .code128: return "code128"
        case .EAN8: return "ean8"
        case .EAN13: return "ean13"
        case .ITF: return "itf"
        case .UPCA: return "upca"
        case .UPCE: return "upce"
        case .aztec: return "aztec"
        case .dataMatrix: return "datamatrix"
        case .PDF417: return "pdf417"
        case .qrCode: return "qrcode"
        default: return "unknown"
        }
    }

    private func valueTypeToString(_ valueType: BarcodeValueType) -> String {
        switch valueType {
        case .text: return "text"
        case .URL: return "url"
        case .email: return "email"
        case .phone: return "phone"
        case .SMS: return "sms"
        case .wiFi: return "wifi"
        case .geographicCoordinates: return "geo"
        case .contactInfo: return "contact"
        case .calendarEvent: return "calendarEvent"
        case .driversLicense: return "driverLicense"
        default: return "unknown"
        }
    }

    // MARK: - Barcode Processing

    private func processBarcodes(_ barcodes: [Barcode], offsetX: CGFloat = 0, offsetY: CGFloat = 0) -> [[String: Any]] {
        var result: [[String: Any]] = []
        for barcode in barcodes {
            let processed = processBarcode(barcode, offsetX: offsetX, offsetY: offsetY)
            result.append(processed)
        }
        return result
    }

    private func processBarcode(_ barcode: Barcode, offsetX: CGFloat, offsetY: CGFloat) -> [String: Any] {
        var dict: [String: Any] = [:]

        dict["rawValue"] = barcode.rawValue ?? ""
        dict["displayValue"] = barcode.displayValue ?? ""
        dict["format"] = barcodeFormatToString(barcode.format)
        dict["valueType"] = valueTypeToString(barcode.valueType)
        dict["frame"] = processRect(barcode.frame, offsetX: offsetX, offsetY: offsetY)
        dict["cornerPoints"] = processCornerPoints(barcode.cornerPoints, offsetX: offsetX, offsetY: offsetY)

        addStructuredData(to: &dict, barcode: barcode)

        return dict
    }

    private func addStructuredData(to dict: inout [String: Any], barcode: Barcode) {
        switch barcode.valueType {
        case .wiFi:
            if let wifi = barcode.wifi {
                dict["wifi"] = processWifi(wifi)
            }
        case .URL:
            if let url = barcode.url?.url {
                dict["url"] = url
            }
        case .email:
            if let email = barcode.email?.address {
                dict["email"] = email
            }
        case .phone:
            if let phone = barcode.phone?.number {
                dict["phone"] = phone
            }
        case .SMS:
            if let sms = barcode.sms {
                dict["sms"] = processSms(sms)
            }
        case .geographicCoordinates:
            if let geo = barcode.geoPoint {
                dict["geo"] = processGeo(geo)
            }
        case .contactInfo:
            if let contact = barcode.contactInfo {
                dict["contact"] = processContact(contact)
            }
        case .calendarEvent:
            if let event = barcode.calendarEvent {
                dict["calendarEvent"] = processCalendarEvent(event)
            }
        case .driversLicense:
            if let license = barcode.driverLicense {
                dict["driverLicense"] = processDriverLicense(license)
            }
        default:
            break
        }
    }

    private func processWifi(_ wifi: BarcodeWifi) -> [String: Any] {
        var dict: [String: Any] = [:]
        dict["ssid"] = wifi.ssid ?? ""
        dict["password"] = wifi.password ?? ""

        let encryptionType: String
        switch wifi.type {
        case .open: encryptionType = "open"
        case .WPA: encryptionType = "wpa"
        case .WEP: encryptionType = "wep"
        default: encryptionType = "unknown"
        }
        dict["encryptionType"] = encryptionType

        return dict
    }

    private func processSms(_ sms: BarcodeSMS) -> [String: Any] {
        var dict: [String: Any] = [:]
        dict["phoneNumber"] = sms.phoneNumber ?? ""
        dict["message"] = sms.message ?? ""
        return dict
    }

    private func processGeo(_ geo: BarcodeGeoPoint) -> [String: Any] {
        var dict: [String: Any] = [:]
        dict["latitude"] = geo.latitude
        dict["longitude"] = geo.longitude
        return dict
    }

    private func processContact(_ contact: BarcodeContactInfo) -> [String: Any] {
        var dict: [String: Any] = [:]

        if let name = contact.name {
            let firstName = name.first ?? ""
            let lastName = name.last ?? ""
            let fullName = "\(firstName) \(lastName)".trimmingCharacters(in: .whitespaces)
            if !fullName.isEmpty {
                dict["name"] = fullName
            }
        }

        if let organization = contact.organization, !organization.isEmpty {
            dict["organization"] = organization
        }

        if let phones = contact.phones, !phones.isEmpty {
            dict["phones"] = phones.compactMap { $0.number }
        }

        if let emails = contact.emails, !emails.isEmpty {
            dict["emails"] = emails.compactMap { $0.address }
        }

        if let urls = contact.urls, !urls.isEmpty {
            dict["urls"] = urls
        }

        if let addresses = contact.addresses, !addresses.isEmpty {
            dict["addresses"] = addresses.compactMap { $0.addressLines?.joined(separator: ", ") }
        }

        return dict
    }

    // Reusable DateFormatter to avoid expensive allocations per frame
    private static let calendarEventDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd'T'HHmmss'Z'"
        formatter.timeZone = TimeZone(abbreviation: "UTC")
        return formatter
    }()

    private func processCalendarEvent(_ event: BarcodeCalendarEvent) -> [String: Any] {
        var dict: [String: Any] = [:]
        dict["summary"] = event.summary ?? ""
        dict["description"] = event.eventDescription ?? ""
        dict["location"] = event.location ?? ""

        if let start = event.start {
            dict["start"] = Self.calendarEventDateFormatter.string(from: start)
        }
        if let end = event.end {
            dict["end"] = Self.calendarEventDateFormatter.string(from: end)
        }

        return dict
    }

    private func processDriverLicense(_ license: BarcodeDriverLicense) -> [String: Any] {
        var dict: [String: Any] = [:]
        dict["firstName"] = license.firstName ?? ""
        dict["lastName"] = license.lastName ?? ""
        dict["middleName"] = license.middleName ?? ""
        dict["gender"] = license.gender ?? ""
        dict["addressStreet"] = license.addressStreet ?? ""
        dict["addressCity"] = license.addressCity ?? ""
        dict["addressState"] = license.addressState ?? ""
        dict["addressZip"] = license.addressZip ?? ""
        dict["licenseNumber"] = license.licenseNumber ?? ""
        dict["birthDate"] = license.birthDate ?? ""
        dict["issuingCountry"] = license.issuingCountry ?? ""
        return dict
    }

    // MARK: - Geometry Processing

    private func processRect(_ rect: CGRect, offsetX: CGFloat = 0, offsetY: CGFloat = 0) -> [String: CGFloat] {
        var dict: [String: CGFloat] = [:]
        dict["x"] = rect.midX + offsetX
        dict["y"] = rect.midY + offsetY
        dict["width"] = rect.width
        dict["height"] = rect.height
        return dict
    }

    private func processCornerPoints(_ cornerPoints: [NSValue]?, offsetX: CGFloat = 0, offsetY: CGFloat = 0) -> [[String: Int]] {
        guard let cornerPoints = cornerPoints else { return [] }
        var result: [[String: Int]] = []
        for pointValue in cornerPoints {
            let point = pointValue.cgPointValue
            var pointDict: [String: Int] = [:]
            pointDict["x"] = Int(point.x + offsetX)
            pointDict["y"] = Int(point.y + offsetY)
            result.append(pointDict)
        }
        return result
    }
}
